### PR TITLE
Customizable Cast Receiver

### PR DESCRIPTION
**Pull request by crobibero**
https://github.com/jellyfin/jellyfin-meta/issues/45

Up for initial review. Doesn't use the existing "ChromecastVersion" preference.

### CODE REVIEW

File: Jellyfin.Data/Entities/User.cs [hunk 1]
```diff
@@ -288,6 +288,11 @@ public User(string username, string authenticationProviderId, string passwordRes
         /// </summary>
         public SyncPlayUserAccessType SyncPlayAccess { get; set; }
 
+        /// <summary>
+        /// Gets or sets the cast receiver id.
+        /// </summary>
+        public string? CastReceiverId { get; set; }
```
Comment (crobibero) [inline Jellyfin.Data/Entities/User.cs:294]: TODO set column size limit

File: MediaBrowser.Model/Configuration/UserConfiguration.cs [hunk 2]
```diff
@@ -69,5 +69,10 @@ public UserConfiguration()
         public bool RememberSubtitleSelections { get; set; }
 
         public bool EnableNextEpisodeAutoPlay { get; set; }
+
+        /// <summary>
+        /// Gets or sets the id of the selected cast receiver.
+        /// </summary>
+        public string? CastReceiverId { get; set; }
```
[the remark this answers is not in the document]
Comment (crobibero) [reply MediaBrowser.Model/Configuration/UserConfiguration.cs:76]: Current implementation- yes

Right now if the user hasn't configured a receiver then the server returns the first in the server config

File: Jellyfin.Data/Entities/User.cs [hunk 3]
```diff
@@ -288,6 +288,12 @@ public User(string username, string authenticationProviderId, string passwordRes
         /// </summary>
         public SyncPlayUserAccessType SyncPlayAccess { get; set; }
 
+        /// <summary>
+        /// Gets or sets the cast receiver id.
+        /// </summary>
+        [StringLength(32)]
```
[the remark this answers is not in the document]
Comment (crobibero) [reply Jellyfin.Data/Entities/User.cs:294]: I looked quickly to see if there was a length specified but I didn't find one, is it documented anywhere?

[the remark this answers is not in the document]
Comment (nielsvanvelzen) [reply Jellyfin.Data/Entities/User.cs:294]: I don't think it is documented but all the ids I've seen in the wild are 8 characters. But if that's not documented we may want to be safe and use 32.

File: MediaBrowser.Model/Configuration/UserConfiguration.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (nielsvanvelzen) [reply MediaBrowser.Model/Configuration/UserConfiguration.cs:76]: Ok that's probably fine. We'll need to check this client side and disable casting if the value is not set. Perhaps we could use this to disable casting for users in the future (although I don't see much of a point in doing so).

**Review by nielsvanvelzen — APPROVED** [review]
✔️ API changes

**Review by Shadowghost — APPROVED** [review]

**Review by JPVenson — APPROVED** [review]

**Review by Bond-009 — APPROVED** [review]

File: Jellyfin.Server.Implementations/Users/UserManager.cs [hunk 4]
```diff
@@ -315,7 +320,11 @@ public UserDto GetUserDto(User user, string? remoteEndPoint = null)
                     OrderedViews = user.GetPreferenceValues<Guid>(PreferenceKind.OrderedViews),
                     GroupedFolders = user.GetPreferenceValues<Guid>(PreferenceKind.GroupedFolders),
                     MyMediaExcludes = user.GetPreferenceValues<Guid>(PreferenceKind.MyMediaExcludes),
-                    LatestItemsExcludes = user.GetPreferenceValues<Guid>(PreferenceKind.LatestItemExcludes)
+                    LatestItemsExcludes = user.GetPreferenceValues<Guid>(PreferenceKind.LatestItemExcludes),
+                    CastReceiverId = string.IsNullOrEmpty(user.CastReceiverId)
+                        ? _serverConfigurationManager.Configuration.CastReceiverApplications.FirstOrDefault()?.Id
```
Comment (cvium) [inline Jellyfin.Server.Implementations/Users/UserManager.cs:325]: Should probably put `_serverConfigurationManager.Configuration.CastReceiverApplications` in a variable. Pretty long lines.

**Pull request merged**